diff --git a/bazel/dependency_imports.bzl b/bazel/dependency_imports.bzl
index b8c8df4..d6a09a4 100644
--- a/bazel/dependency_imports.bzl
+++ b/bazel/dependency_imports.bzl
@@ -1,6 +1,6 @@
-load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
+#load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
 load("@com_google_protobuf//:protobuf_deps.bzl", "protobuf_deps")
-load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")
+#load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")
 load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")
 load("@rules_python//python:pip.bzl", "pip_install")

@@ -13,11 +13,11 @@ def _pgv_pip_dependencies():
     )

 def _pgv_go_dependencies():
-    go_rules_dependencies()
-    go_register_toolchains(
-        version = "1.15.6",
-    )
-    gazelle_dependencies()
+    #go_rules_dependencies()
+    #go_register_toolchains(
+    #    version = "1.15.6",
+    #)
+    #gazelle_dependencies()

 def pgv_dependency_imports():
     # Import @com_google_protobuf's dependencies.
@@ -27,7 +27,7 @@ def pgv_dependency_imports():
     _pgv_pip_dependencies()

     # Import rules for the Go compiler.
-    _pgv_go_dependencies()
+    #_pgv_go_dependencies()

     # Setup rules_proto.
     rules_proto_dependencies()
diff --git a/validate/BUILD b/validate/BUILD
index 5040d64..94d5e94 100644
--- a/validate/BUILD
+++ b/validate/BUILD
@@ -1,9 +1,9 @@
 load("@com_google_protobuf//:protobuf.bzl", "py_proto_library")
-load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
+#load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 load("@rules_cc//cc:defs.bzl", "cc_library", "cc_proto_library")
 load("@rules_java//java:defs.bzl", "java_proto_library")
 load("@rules_proto//proto:defs.bzl", "proto_library")
-load("@io_bazel_rules_go//go:def.bzl", "go_library")
+#load("@io_bazel_rules_go//go:def.bzl", "go_library")

 package(
     default_visibility =
@@ -31,22 +31,22 @@ py_proto_library(
     deps = ["@com_google_protobuf//:protobuf_python"],
 )

-go_proto_library(
-    name = "validate_go_proto",
-    importpath = "github.com/envoyproxy/protoc-gen-validate/validate",
-    proto = ":validate_proto",
-)
+#go_proto_library(
+#    name = "validate_go_proto",
+#    importpath = "github.com/envoyproxy/protoc-gen-validate/validate",
+#    proto = ":validate_proto",
+#)

 cc_library(
     name = "cc_validate",
     hdrs = ["validate.h"],
 )

-go_library(
-    name = "validate_go",
-    embed = [":validate_go_proto"],
-    importpath = "github.com/envoyproxy/protoc-gen-validate/validate",
-)
+#go_library(
+#    name = "validate_go",
+#    embed = [":validate_go_proto"],
+#    importpath = "github.com/envoyproxy/protoc-gen-validate/validate",
+#)

 java_proto_library(
     name = "validate_java",