diff --git a/bazel/upb_proto_library.bzl b/bazel/upb_proto_library.bzl
--- a/bazel/build_defs.bzl
+++ b/bazel/build_defs.bzl
@@ -34,13 +34,13 @@
 _DEFAULT_CPPOPTS.extend([
     "-Wextra",
     # "-Wshorten-64-to-32",  # not in GCC (and my Kokoro images doesn't have Clang)
-    "-Werror",
+    # "-Werror",
     "-Wno-long-long",
 ])
 _DEFAULT_COPTS.extend([
     "-std=c99",
     "-pedantic",
-    "-Werror=pedantic",
+    # "-Werror=pedantic",
     "-Wall",
     "-Wstrict-prototypes",
     # GCC (at least) emits spurious warnings for this that cannot be fixed
