diff --git a/java/util/BUILD b/java/util/BUILD
index ee6ddeaf1..54e20ea24 100644
--- a/java/util/BUILD
+++ b/java/util/BUILD
@@ -13,11 +13,11 @@ java_library(
     deps = [
         "//java/core",
         "//java/lite",
-        "@maven//:com_google_code_findbugs_jsr305",
-        "@maven//:com_google_code_gson_gson",
-        "@maven//:com_google_errorprone_error_prone_annotations",
-        "@maven//:com_google_guava_guava",
-        "@maven//:com_google_j2objc_j2objc_annotations",
+        "//external:jsr305",
+        "//external:gson",
+        "//external:error_prone_annotations",
+        "//external:guava",
+        "//external:j2objc_annotations",
     ],
 )
 
