diff --git a/bazel/common/BUILD b/bazel/common/BUILD
index c79d22280..9ab40ff05 100644
--- a/bazel/common/BUILD
+++ b/bazel/common/BUILD
@@ -22,7 +22,8 @@ bzl_library(
     ],
     visibility = ["//visibility:public"],
     deps = [
-        "//bazel/private:native_bzl",
+        "//bazel/private:proto_info_bzl",
+        "@proto_bazel_features//:features",
     ],
 )
 
diff --git a/bazel/private/BUILD b/bazel/private/BUILD
index 9b04c1f86..8e5207f3a 100644
--- a/bazel/private/BUILD
+++ b/bazel/private/BUILD
@@ -108,6 +108,14 @@ bzl_library(
     ],
 )
 
+bzl_library(
+    name = "proto_info_bzl",
+    srcs = [
+        "proto_info.bzl",
+    ],
+    visibility = ["//bazel:__subpackages__"],
+)
+
 native_bool_flag(
     name = "experimental_proto_descriptor_sets_include_source_info",
     flag = "experimental_proto_descriptor_sets_include_source_info",
