diff --git a/rules/rbe_repo/build_gen.bzl b/rules/rbe_repo/build_gen.bzl
index 833fc1f..bdce12c 100644
--- a/rules/rbe_repo/build_gen.bzl
+++ b/rules/rbe_repo/build_gen.bzl
@@ -90,7 +90,7 @@ def create_java_runtime(ctx, java_home, java_version):
       java_version: the Java runtime release version
     """
     bazel_version = tuple([int(n) for n in ctx.attr.bazel_version.split(".")])
-    if bazel_version > (4, 0, 0) or not native.bazel_version:
+    if bazel_version > (5, 0, 0) or not native.bazel_version:
         template = ctx.path(Label("@bazel_toolchains//rules/rbe_repo:BUILD.local_java_runtime.tpl"))
     else:
         template = ctx.path(Label("@bazel_toolchains//rules/rbe_repo:BUILD.java.tpl"))
