diff --git a/lib/extensions.bzl b/lib/extensions.bzl
index cae31c0..5909db8 100644
--- a/lib/extensions.bzl
+++ b/lib/extensions.bzl
@@ -87,8 +87,9 @@ def _toolchains_extension_impl(mctx):
         get_version_fn = lambda attr: attr.core_version,
     )
 
-    if bazel_features.external_deps.extension_metadata_has_reproducible:
-        return mctx.extension_metadata(reproducible = True)
+    # Do not return `reproducible`: it complicates bazel's distfile bootstrap.
+    # if bazel_features.external_deps.extension_metadata_has_reproducible:
+    #     return mctx.extension_metadata(reproducible = True)
 
     return mctx.extension_metadata()
 
