diff --git a/internal/native_image/builder.bzl b/internal/native_image/builder.bzl
index ee4a38f..a7f6cb2 100644
--- a/internal/native_image/builder.bzl
+++ b/internal/native_image/builder.bzl
@@ -236 +236 @@ def assemble_native_build_options(
-        args.add(arg)
+        args.add(ctx.expand_location(arg, targets = ctx.attr.additional_inputs))
diff --git a/internal/native_image/classic.bzl b/internal/native_image/classic.bzl
index 907fdc9..ff6f211 100644
--- a/internal/native_image/classic.bzl
+++ b/internal/native_image/classic.bzl
@@ -28 +28 @@ def _graal_binary_classic_implementation(ctx):
-    direct_inputs = []
+    direct_inputs = ctx.files.additional_inputs[:]
diff --git a/internal/native_image/common.bzl b/internal/native_image/common.bzl
index 5e43aad..48ae32e 100644
--- a/internal/native_image/common.bzl
+++ b/internal/native_image/common.bzl
@@ -102,0 +103,3 @@ _NATIVE_IMAGE_ATTRS = {
+    "additional_inputs": attr.label_list(
+        allow_files = True,
+    ),
diff --git a/internal/native_image/rules.bzl b/internal/native_image/rules.bzl
index 0a1b65a..7879960 100644
--- a/internal/native_image/rules.bzl
+++ b/internal/native_image/rules.bzl
@@ -50 +50 @@ def _graal_binary_implementation(ctx):
-    direct_inputs = []
+    direct_inputs = ctx.files.additional_inputs[:]
